vodesfunc/descale.py

```diff
@@ -254,6 +272,26 @@ def crossconv_shift_calc_irregular(clip: vs.VideoNode, native_height: int) -> fl
         """Calculate the shift for an irregular cross-conversion."""
         return 0.25 / (clip.height / native_height)
 
+    def _set_bord_mask(self, clip: vs.VideoNode) -> vs.VideoNode:
+        """Set an attribute for a border mask for border_handling."""
+        if (bord_rad := self.border_radius) is None:
+            try:
+                bord_rad = self.kernel.kernel_radius
+            except (AttributeError, NotImplementedError):
+                bord_rad = 2
+
+        # Expands border mask radii based on the scale factor. Only really relevant for factors ~2.0.
+        # Reduce by a bit more than 1 to stop frac ~720p resolutions from needlessly increasing the radius.
+        bord_rad += round(clip.height / self.height - 1 - 0.025)
```

Comment:
The exact formula for the number of dirty lines at the top of an image is:
`1 + floor( ( (base_height - 1) / 2 + kernel_support - src_top - 1 ) * target_height / src_height + (target_height - 1) / 2 ))`
(to get the number of dirty lines at the bottom, replace `src_top` with `base_height - src_height - src_top`).

To see why this is true: The list of points where an interpolated image is sampled during resampling is
`[src_top + (n - (target_height - 1) / 2) * src_height / target_height + (base_height - 1) / 2 for n in range(target_height)]`
The dirty lines at the top are the points whose y-coordinate is larger than `-1 + kernel_support`. You then get the above value by solving the inequality for `n`.

Reply:
>to get the number of dirty lines at the bottom, replace `src_top` with `base_height - src_height - src_top`

How about on the left/right? I'd like to be able to handle all cases in one go if possible if only for my own sanity.

Reply:
Same thing, just replace everything with the horizontal versions. `height` -> `width` and `top` -> `left`.

Reply:
This has now been implemented. Thanks!

Reply:
It  turned out this was incorrect for fractional upscales: `base_height` should just be `src_height` in the formula. Also, using `ceil` instead of `1 + floor` is conceptually cleaner (and more correct in edge cases where samples align exactly). So we get
`[src_top + (n - (target_height - 1) / 2) * src_height / target_height + (src_height - 1) / 2 for n in range(target_height)]`
for the list of sample points, and
`ceil( ( kernel_support - 1 - src_top - (src_height - 1) / 2 ) * target_height / src_height + (target_height - 1) / 2 ) )`
for the number of dirty lines.